src/coreclr/src/gc/unix/cgroup.cpp

```diff
@@ -59,96 +67,116 @@ class CGroup
 
     static bool GetPhysicalMemoryLimit(uint64_t *val)
     {
-        char *mem_limit_filename = nullptr;
-        bool result = false;
-
-        if (s_memory_cgroup_path == nullptr)
-            return result;
-
-        size_t len = strlen(s_memory_cgroup_path);
-        len += strlen(MEM_LIMIT_FILENAME);
-        mem_limit_filename = (char*)malloc(len+1);
-        if (mem_limit_filename == nullptr)
-            return result;
-
-        strcpy(mem_limit_filename, s_memory_cgroup_path);
-        strcat(mem_limit_filename, MEM_LIMIT_FILENAME);
-        result = ReadMemoryValueFromFile(mem_limit_filename, val);
-        free(mem_limit_filename);
-        return result;
+        if (s_cgroup_version == 0)
+            return false;
+        else if (s_cgroup_version == 1)
+            return GetCGroupMemoryLimit(val, CGROUP1_MEMORY_LIMIT_FILENAME);
+        else if (s_cgroup_version == 2)
+            return GetCGroupMemoryLimit(val, CGROUP2_MEMORY_LIMIT_FILENAME);
+        else
+        {
+            assert(!"Unknown cgroup version.");
+            return false;
+        }
     }
 
     static bool GetPhysicalMemoryUsage(size_t *val)
     {
-        char *mem_usage_filename = nullptr;
-        bool result = false;
-        uint64_t temp = 0;
+        if (s_cgroup_version == 0)
+            return false;
+        else if (s_cgroup_version == 1)
+            return GetCGroupMemoryUsage(val, CGROUP1_MEMORY_USAGE_FILENAME);
+        else if (s_cgroup_version == 2)
+            return GetCGroupMemoryUsage(val, CGROUP2_MEMORY_USAGE_FILENAME);
+        else
+        {
+            assert(!"Unknown cgroup version.");
+            return false;
+        }
+    }
 
-        if (s_memory_cgroup_path == nullptr)
-            return result;
+    static bool GetCpuLimit(uint32_t *val)
+    {
+        if (s_cgroup_version == 0)
+            return false;
+        else if (s_cgroup_version == 1)
+            return GetCGroup1CpuLimit(val);
+        else if (s_cgroup_version == 2)
+            return GetCGroup2CpuLimit(val);
+        else
+        {
+            assert(!"Unknown cgroup version.");
+            return false;
+        }
+    }
 
-        size_t len = strlen(s_memory_cgroup_path);
-        len += strlen(MEM_USAGE_FILENAME);
-        mem_usage_filename = (char*)malloc(len+1);
-        if (mem_usage_filename == nullptr)
-            return result;
+private:
+    static int FindCGroupVersion()
+    {
+        int version = 0;
+        char *line = nullptr;
+        size_t lineLen = 0;
+        char *filesystemType = nullptr;
 
-        strcpy(mem_usage_filename, s_memory_cgroup_path);
-        strcat(mem_usage_filename, MEM_USAGE_FILENAME);
-        result = ReadMemoryValueFromFile(mem_usage_filename, &temp);
-        if (result)
+        FILE *mountinfofile = fopen(PROC_MOUNTINFO_FILENAME, "r");
+        if (mountinfofile == nullptr)
+            goto done;
```

Comment:
You can just `return version` here and not have to worry about initializing `line`, or checking if `mountinfofile` is `NULL` before calling `fclose()` on it.  (Same for all error paths.)

Reply:
Sorry, I am probably being thick, but what other error paths in this code are safe to modify as you suggest? Dont we have to `flcose()` from all other error paths?